pkg/dockerregistry/server/writelimiter/semaphore.go

```diff
@@ -0,0 +1,24 @@
+package writelimiter
+
+type semaphore struct {
+	c chan struct{}
+}
+
+func newSemaphore(count int) *semaphore {
+	return &semaphore{
+		c: make(chan struct{}, count),
+	}
+}
+
+func (s *semaphore) Up() {
+	<-s.c
+}
+
+func (s *semaphore) TryDown() bool {
```

Comment:
Sorry, deadline to wait to get a chance to run.  Once we pass the deadline, give up rather than keep waiting.  If we get a chance to run before the deadline, don't wait.

If we have to we can implement slow writer and reader protection higher up the stack in the blob writer (i'm actually fairly worried about that in general, we just need to get past this hump)